File: src/Atlasd/Battlenet/Protocols/Game/Messages/SID_STARTADVEX2.cs
using Atlasd.Battlenet.Exceptions;
using Atlasd.Daemon;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace Atlasd.Battlenet.Protocols.Game.Messages
{
    class SID_STARTADVEX2 : Message
    {
        public enum Statuses : UInt32
        {
            Failed = 0,
            Success = 1,
        };

        public SID_STARTADVEX2()
        {
            Id = (byte)MessageIds.SID_STARTADVEX2;
            Buffer = new byte[0];
        }

        public SID_STARTADVEX2(byte[] buffer)
        {
            Id = (byte)MessageIds.SID_STARTADVEX2;
            Buffer = buffer;
        }

        public override bool Invoke(MessageContext context)
        {
            switch (context.Direction)
            {
                case MessageDirection.ClientToServer:
                    {
                        Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)");

                        /**
                         * (UINT32) Status
                         * (UINT32) Unknown [editor's note: probably elapsed time]
                         * (UINT16) Game Type
                         * (UINT16) Unknown [editor's note: probably sub game type]
                         * (UINT32) Unknown [editor's note: probably viewing filter]
                         * (UINT32) Port
                         * (STRING) Game name
                         * (STRING) Game password
                         * (STRING) Game stats - Flags, Creator, Statstring
                         */

                        if (Buffer.Length < 23)
                            throw new GameProtocolViolationException(context.Client, $"{MessageName(Id)} buffer must be at least 23 bytes");

                        if (context.Client.GameState.ActiveAccount == null)
                            throw new GameProtocolViolationException(context.Client, $"{MessageName(Id)} was received before logon");

                        using var m = new MemoryStream(Buffer);
                        using var r = new BinaryReader(m);

                        var gameState = r.ReadUInt32();
                        var gameElapsedTime = r.ReadUInt32();
                        var gameType = r.ReadUInt16();
                        var subGameType = r.ReadUInt16();
                        var viewingFilter = r.ReadUInt32();
                        var portNumber = r.ReadUInt32();
                        var gameName = r.ReadByteString();
                        var gamePassword = r.ReadByteString();
                        var gameStatstring = r.ReadByteString();

                        Statuses status = Statuses.Failed;
                        GameAd gameAd = null;

                        lock (Battlenet.Common.ActiveGameAds)
                        {
                            foreach (GameAd _ad in Battlenet.Common.ActiveGameAds)
                            {
                                if (_ad.Name.SequenceEqual(gameName))
                                {
                                    gameAd = _ad;
                                    break;
                                }
                            }

                            if (gameAd == null)
                            {
                                gameAd = new GameAd(context.Client.GameState, gameName, gamePassword, gameStatstring, 6112, (GameAd.GameTypes)gameType, subGameType, context.Client.GameState.Version.VersionByte);
                                gameAd.AddClient(context.Client.GameState);
                                Battlenet.Common.ActiveGameAds.Add(gameAd);
                            }
                        }

                        if (gameAd.HasClient(context.Client.GameState)) context.Client.GameState.GameAd = gameAd;

                        bool gameAdOwner = gameAd != null && gameAd.Owner == context.Client.GameState;
                        if (!gameAdOwner)
                        {
                            status = Statuses.Failed;
                        }
                        else
                        {
                            status = Statuses.Success;
                            gameAd.SetActiveStateFlags((GameAd.StateFlags)gameState);
                            gameAd.SetElapsedTime(gameElapsedTime);
                            gameAd.SetGameType((GameAd.GameTypes)gameType);
                            gameAd.SetName(gameName);
                            gameAd.SetPassword(gamePassword);
                            gameAd.SetPort(6112);
                            gameAd.SetStatstring(gameStatstring);
                        }

                        return new SID_STARTADVEX2().Invoke(new MessageContext(context.Client, MessageDirection.ServerToClient, new Dictionary<string, dynamic>(){{ "status", status }}));
                    }
                case MessageDirection.ServerToClient:
                    {
                        var status = context.Arguments.ContainsKey("status") ? (Statuses)context.Arguments["status"] : Statuses.Failed;

                        /**
                         * (UINT32) Status
                         */

                        Buffer = new byte[4];

                        using var m = new MemoryStream(Buffer);
                        using var w = new BinaryWriter(m);

                        w.Write((UInt32)status);

                        Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)");
                        context.Client.Send(ToByteArray(context.Client.ProtocolType));
                        return true;
                    }
            }

            return false;
        }
    }
}

File: src/AtlasV/Battlenet/Protocols/Game/Messages/SID_STARTADVEX2.vb
Imports AtlasV.Battlenet.Exceptions
Imports AtlasV.Daemon
Imports System
Imports System.IO

Namespace AtlasV.Battlenet.Protocols.Game.Messages
    Class SID_STARTADVEX2
        Inherits Message

        Public Sub New()
            Id = CByte(MessageIds.SID_STARTADVEX2)
            Buffer = Array.Empty(Of Byte)()
        End Sub

        Public Sub New(ByVal varBuffer As Byte())
            Id = CByte(MessageIds.SID_STARTADVEX2)
            Buffer = varBuffer
        End Sub

        Public Overrides Function Invoke(ByVal context As MessageContext) As Boolean
            Select Case context.Direction
                Case MessageDirection.ClientToServer
                    Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)")
                    If Buffer.Length < 23 Then Throw New GameProtocolViolationException(context.Client, $"{MessageName(Id)} buffer must be at least 23 bytes")
                    If context.Client.GameState.ActiveAccount Is Nothing Then Throw New GameProtocolViolationException(context.Client, $"{MessageName(Id)} was received before logon")

                    Dim gameState, gameElapsedTime, viewingFilter, portNumber As UInt32
                    Dim gameType, subGameType As UInt16
                    Dim gameName(), gamePassword(), gameStatstring() As Byte
                    Using m = New MemoryStream(Buffer)
                        Using r = New BinaryReader(m)
                            gameState = r.ReadUInt32()
                            gameElapsedTime = r.ReadUInt32()
                            gameType = r.ReadUInt16()
                            subGameType = r.ReadUInt16()
                            viewingFilter = r.ReadUInt32()
                            portNumber = r.ReadUInt32()
                            gameName = r.ReadByteString()
                            gamePassword = r.ReadByteString()
                            gameStatstring = r.ReadByteString()
                        End Using
                    End Using

                    Dim gameAds = Battlenet.Common.ActiveGameAds.ToArray()
                    Dim gameAd As GameAd = Nothing

                    For Each _pair In gameAds

                        If Object.Equals(_pair.Value.Client, context.Client.GameState) = True Then
                            gameAd = _pair.Value
                            Exit For
                        End If
                    Next

                    If gameAd Is Nothing Then
                        gameAd = New GameAd(context.Client.GameState, gameName, gamePassword, gameStatstring, 6112, CType(gameType, GameAd.GameTypes), subGameType, context.Client.GameState.Version.VersionByte)
                        context.Client.GameState.GameAd = gameAd
                        Battlenet.Common.ActiveGameAds.TryAdd(gameName, gameAd)
                    End If

                    gameAd.SetActiveStateFlags(CType(gameState, GameAd.StateFlags))
                    gameAd.SetElapsedTime(gameElapsedTime)
                    gameAd.SetGameType(CType(gameType, GameAd.GameTypes))
                    gameAd.SetName(gameName)
                    gameAd.SetPassword(gamePassword)
                    gameAd.SetPort(portNumber)
                    gameAd.SetStatstring(gameStatstring)
                    Return New SID_STARTADVEX2().Invoke(New MessageContext(context.Client, MessageDirection.ServerToClient))
                Case MessageDirection.ServerToClient
                    Buffer = New Byte(3) {}

                    Using m = New MemoryStream(Buffer)
                        Using w = New BinaryWriter(m)
                            w.Write(CUInt(1))
                        End Using
                    End Using

                    Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)")
                    context.Client.Send(ToByteArray(context.Client.ProtocolType))
                    Return True
            End Select

            Return False
        End Function
    End Class
End Namespace
